### PR TITLE
ddl: fix can't pause add index in dist task mode

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/46892, ref #46258 

Problem Summary:
add index running on dist task framework can't pause and resume properly.
### What is changed and how it works?
pause and resume the dist task.
### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -2043,30 +2045,65 @@ func (w *worker) executeDistGlobalTask(reorgInfo *reorgInfo) error {
...
+		metaData, err := json.Marshal(taskMeta)
+		if err != nil {
+			return err
+		}
+
+		g.Go(func() error {
+			defer close(done)
+			err := handle.SubmitAndRunGlobalTask(ctx, taskKey, taskType, distPhysicalTableConcurrency, metaData)
+			if w.isReorgPaused(reorgInfo.Job.ID) {
+				logutil.BgLogger().Warn("job paused by user", zap.String("category", "ddl"))
```
Comment (zimulala) [inline ddl/index.go:2101]: Do we need to print the `err` message？ Maybe it‘s not nil.

File: ddl/index.go [hunk 2]
```diff
@@ -2043,30 +2045,65 @@ func (w *worker) executeDistGlobalTask(reorgInfo *reorgInfo) error {
...
+				func(ctx context.Context) (bool, error) {
+					return true, handle.ResumeTask(taskKey)
+				},
+			)
+			if err != nil {
+				return err
+			}
+			err = handle.WaitGlobalTask(ctx, task)
+			if w.isReorgPaused(reorgInfo.Job.ID) {
+				logutil.BgLogger().Warn("job paused by user", zap.String("category", "ddl"))
```
Comment (zimulala) [inline ddl/index.go:2074]: Ditto.

File: ddl/index.go [hunk 3]
```diff
@@ -2043,30 +2045,65 @@ func (w *worker) executeDistGlobalTask(reorgInfo *reorgInfo) error {
...
+			err := handle.RunWithRetry(ctx, dispatcher.RetrySQLTimes, backoffer, logutil.BgLogger(),
+				func(ctx context.Context) (bool, error) {
+					return true, handle.ResumeTask(taskKey)
+				},
+			)
+			if err != nil {
+				return err
+			}
+			err = handle.WaitGlobalTask(ctx, task)
+			if w.isReorgPaused(reorgInfo.Job.ID) {
```
Comment (zimulala) [inline ddl/index.go:2073]: How about the case: we are pausing when the task has `TaskStateSucceed`.

File: disttask/framework/handle/handle.go [hunk 4]
```diff
@@ -90,6 +89,9 @@ func WaitGlobalTask(ctx context.Context, globalTask *proto.Task) error {
 			case proto.TaskStateReverted:
 				logutil.BgLogger().Error("global task reverted", zap.Int64("task-id", globalTask.ID), zap.Error(found.Error))
 				return found.Error
+			case proto.TaskStatePaused:
```
Comment (Benjamin2037) [inline disttask/framework/handle/handle.go:92]: If task in paused status，will the log print too much due to dispatche manger period check？

File: ddl/index.go (hunk 3 above)
Comment (ywqzzy) [reply ddl/index.go:2073]: > TaskStateSucceed

These transform is forbidden by `func (d *BaseDispatcher) VerifyTaskStateTransform(from, to string) bool`

File: disttask/framework/handle/handle.go (hunk 4 above)
Comment (ywqzzy) [reply disttask/framework/handle/handle.go:92]: > If task in paused status，will the log print too much due to dispatche manger period check？

No, dispatcher exits.

File: ddl/index.go (hunk 3 above)
Comment (zimulala) [reply ddl/index.go:2073]: I mean there is a time difference between the determination of `WaitGlobalTask ` and `isReorgPaused` functions
* The global task is finished (real situation)
* We receive a pause DDL operation which means `w.isReorgPaused(reorgInfo.Job.ID) ` is false (This judgment may be delayed)

File: tests/realtikvtest/addindextest/add_index_test.go [hunk 5]
```diff
@@ -202,3 +207,74 @@ func TestAddIndexDistCancel(t *testing.T) {
...
+		require.Equal(t, 1, len(row))
+		jobID := row[0][0].(string)
+		tk1.MustExec("admin resume ddl jobs " + jobID)
+	}
+
+	ddl.MockDMLExecutionOnTaskFinished = func() {
+		row := tk1.MustQuery("select job_id from mysql.tidb_ddl_job").Rows()
+		require.Equal(t, 1, len(row))
+		jobID := row[0][0].(string)
+		logutil.BgLogger().Info("ywq test pause")
```
Comment (zimulala) [inline tests/realtikvtest/addindextest/add_index_test.go:248]: Do we need to remove this line？

File: tests/realtikvtest/addindextest/add_index_test.go [hunk 6]
```diff
@@ -202,3 +207,74 @@ func TestAddIndexDistCancel(t *testing.T) {
...
+	require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/disttask/framework/dispatcher/mockDMLExecutionOnPausedState"))
+
+	// dist task succeed, job paused and resumed.
+	var hook = &callback.TestDDLCallback{Do: dom}
+	var resumeFunc = func(job *model.Job) {
+		if job.IsPaused() {
+			row := tk1.MustQuery("select job_id from mysql.tidb_ddl_job").Rows()
+			require.Equal(t, 1, len(row))
+			jobID := row[0][0].(string)
+			logutil.BgLogger().Info("ywq test resume")
```
Comment (zimulala) [inline tests/realtikvtest/addindextest/add_index_test.go:269]: Ditto.

File: disttask/framework/dispatcher/dispatcher.go [hunk 7]
```diff
@@ -251,9 +251,17 @@ func (d *BaseDispatcher) onPausing() error {
...
 }
 
+// MockDMLExecutionOnPausedState is used to mock DML execution when tasks paused.
+var MockDMLExecutionOnPausedState func(task *proto.Task)
+
 // handle task in paused state
 func (d *BaseDispatcher) onPaused() error {
 	logutil.Logger(d.logCtx).Info("on paused state", zap.String("state", d.Task.State), zap.Int64("stage", d.Task.Step))
+	failpoint.Inject("mockDMLExecutionOnPausedState", func(val failpoint.Value) {
+		if val.(bool) && MockDMLExecutionOnPausedState != nil {
```
[the remark this answers is not in the document]
Comment (zimulala) [reply disttask/framework/dispatcher/dispatcher.go:261]: Rest LGTM

[the remark this answers is not in the document]
Comment (ywqzzy) [reply disttask/framework/dispatcher/dispatcher.go:261]: > Could we remove `&& MockDMLExecutionOnPausedState != nil`?

I followed the code style of MockDMLExecutionAddIndexSubTaskFinish

[the remark this answers is not in the document]
Comment (zimulala) [reply disttask/framework/dispatcher/dispatcher.go:261]: Could we remove them together? It doesn't feel like it's necessary.

[the remark this answers is not in the document]
Comment (ywqzzy) [reply disttask/framework/dispatcher/dispatcher.go:261]: > Could we remove them together? It doesn't feel like it's necessary.

Ok

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**